packages/core/src/validate/run.ts

```diff
@@ -206,27 +225,159 @@ function* getConstructorErrors(contractDef: ContractDefinition, decodeSrc: SrcDe
+  if (parentNode === undefined || !skipCheck(opcode.kind, parentNode)) {
+    yield* getDirectFunctionOpcodeErrors(contractOrFunctionDef, decodeSrc, opcode, skipInternal);
+  }
+  if (parentNode === undefined || !skipCheckReachable(opcode.kind, parentNode)) {
+    yield* getReferencedFunctionOpcodeErrors(
+      contractOrFunctionDef,
+      deref,
+      decodeSrc,
+      opcode,
+      skipInternal,
+      visitedNodeIds,
+    );
+  }
+}
+
+function* getDirectFunctionOpcodeErrors(
+  contractOrFunctionDef: ContractDefinition | FunctionDefinition,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  skipInternal: boolean,
+) {
+  for (const fnCall of findAll(
+    'FunctionCall',
+    contractOrFunctionDef,
+    node => skipCheck(opcode.kind, node) || skipInternalFunctions(skipInternal, node),
+  )) {
     const fn = fnCall.expression;
-    if (fn.typeDescriptions.typeIdentifier?.match(/^t_function_baredelegatecall_/)) {
+    if (fn.typeDescriptions.typeIdentifier?.match(opcode.pattern)) {
       yield {
-        kind: 'delegatecall',
+        kind: opcode.kind,
         src: decodeSrc(fnCall),
       };
     }
   }
-  for (const fnCall of findAll('FunctionCall', contractDef, node => skipCheck('selfdestruct', node))) {
+}
+
+function* getReferencedFunctionOpcodeErrors(
+  contractOrFunctionDef: ContractDefinition | FunctionDefinition,
+  deref: ASTDereferencer,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  skipInternal: boolean,
+  visitedNodeIds: Set<number>,
+) {
+  for (const fnCall of findAll(
+    'FunctionCall',
+    contractOrFunctionDef,
+    node => skipCheckReachable(opcode.kind, node) || skipInternalFunctions(skipInternal, node),
+  )) {
     const fn = fnCall.expression;
-    if (fn.typeDescriptions.typeIdentifier?.match(/^t_function_selfdestruct_/)) {
-      yield {
-        kind: 'selfdestruct',
-        src: decodeSrc(fnCall),
-      };
+    const fnReference = (fn as any).referencedDeclaration;
+    if (fnReference !== undefined && fnReference > 0) {
+      try {
+        const referencedNode = deref(['FunctionDefinition'], fnReference);
+        if (!visitedNodeIds.has(referencedNode.id)) {
+          visitedNodeIds.add(referencedNode.id);
+          yield* getFunctionOpcodeErrors(referencedNode, deref, decodeSrc, opcode, false, visitedNodeIds);
+        }
+      } catch (e: any) {
+        if (!e.message.includes('No node with id')) {
+          throw e;
+        }
+      }
     }
   }
 }
 
+function* getInheritedContractOpcodeErrors(
+  contractDef: ContractDefinition,
+  deref: ASTDereferencer,
+  decodeSrc: SrcDecoder,
+  opcode: OpcodePattern,
+  visitedNodeIds: Set<number>,
+) {
+  if (!skipCheckReachable(opcode.kind, contractDef)) {
+    for (const base of contractDef.baseContracts) {
+      const referencedContract = deref('ContractDefinition', base.baseName.referencedDeclaration);
+      yield* getContractOpcodeErrors(referencedContract, deref, decodeSrc, opcode, true, visitedNodeIds);
```

Comment:
I'm not sure that we're catching errors in parent constructors if they're internal.

Do we have a test for that? We should test the two different syntaxes for invoking a parent constructor:

```solidity
contract Parent {
    constructor(uint x) internal {
      selfdestruct(msg.sender);
    }
}
contract Child1 is Parent {
    constructor() Parent(0) {}
}
contract Child2 is Parent(0) {}
```

What if theif the parent constructor has no arguments, i.e. it's implicitly invoked?

Reply:
Addressed this in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/702/commits/4aa9b4c457db7df6ddeb4ba70b1dbed5ad7c5b1e by always checking parent constructors even if they are 'internal' functions.  This is because child contracts always call parent constructors, either explicitly or implicitly.